doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md

```diff
@@ -0,0 +1,1069 @@
+4. Define ABNF schema for BGP features used in SONiC
+
+### 1.1.3 Scalability Requirements
+N/A
+
+### 1.1.4 Warm Boot Requirements
+As state and statistics information is retrieved from FRR-BGP on demand there is no Warm Boot specific requirements for this feature.
+
+## 1.2 Design Overview
+SONiC FRR-BGP Extended Unified config and management capability makes use of Management framework to implement the backend and transformer methods to support Open Config data models for BGP and route policy feature. The backend converts the incoming request to Redis ABNF schema format and writes the configuration to Redis DB. Then from DB events, bgpcfgd will configure FRR-BGP using FRR CLI commands.
+It also uses management framework's transformer methods to do syntactic and semantic validation of the requests using ABNF JSON before writing them into the Redis DB.
+
+### 1.2.1 Basic Approach
+
+* This enhancement takes comprehensive approach to support BGP features used in SONiC:
+	* Standard based YANG models and custom YANG models
+	* Open API spec
+	* Industry standard CLI
+	* Config Validation
+* REST server, gNMI server, Transformer methods - all in Go
+* Marshalling and unmarshalling using YGOT
+* Redis updated using CAS(Check-and-Set) trans. (No locking, No rollback)
+* Config Validation by using YANG model from ABNF schema
+
+### 1.2.2 Container
+
+There will be changes in following containers,
+* sonic-mgmt-framework
+* sonic-frr
+
+### 1.2.3 SAI Overview
+N/A - software feature
+
+
+# 2 Functionality
+## 2.1 Target Deployment Use Cases
+Configure and manage FRR-BGP via gNMI, REST and CLI interfaces using SONiC Management Framework.
+
+## 2.2 Functional Description
+Provide GNMI and REST support for config get/set, state get and statistics get,  CLI config and show commands for FRR-BGP features used in SONiC.
+
+
+# 3 Design
+## 3.1 Overview
+The extended unified config and management framework for FRR-BGP in SONiC is represented in below diagram.
+
+![FRR-BGP Unified Mgmt Framework](images/FRR-BGP-Unified-mgmt-frmwrk.png)
+
+1. Transformer common app owns the Open config data models related to BGP (which means no separate app module required for handling BGP open-config and augmented YANG objects).
+
+2. Provide annotations for required objects so that transformer core and common app will take care of handling them.
+
+3. Provide transformer methods as per the annotations defined to take care of model specific logics and validations.
+
+4. Define SONiC YANG and Redis ABNF schema for the supported Open Config BGP models & objects.
+
+5. KLISH CLI and REST clients provide extensive BGP configurations and hence there should not be any need for BGP configurations via vtysh.
+
+6. In bgpcfgd register for Redis DB events for the BGP and other related objects, so as to translate the Redis DB events to FRR-BGP CLI commands to configure FRR-BGP, similarly, separate config daemons can be present to configure individual features like OSPF, BFD..etc
```

Comment:
What if some objects relies on other object. Or changing some objects requires to restart something?

Reply:
We have the cache in bgpcfgd, this will help to retain the information until the dependent data is available, also, it helps to delete the configs upon field delete in BGP tables.